### PR TITLE
Do not optimize append-only operation if normal operation with higher seq# was seen

### PR DESCRIPTION
**Pull request by dnhatn**
When processing an append-only operation, primary knows that operations can only conflict with another instance of the same operation. This is true as the id was freshly generated. However this property doesn't hold for replicas. As soon as an auto-generated ID was indexed into the primary, it can be exposed to a search and users can issue a follow up operation on it. In extremely rare cases, the follow up operation can be arrived and processed on a replica before the original append-only request. In this case we can't simply proceed with the append-only request and blindly add it to the index without consulting the version map. The following scenario can cause difference between primary and replica.

1) Primary indexes an auto-gen-id doc. (id=X, v=1, s#=20)
2) A refresh cycle happens on primary
3) The new doc is picked up and modified - say by a delete by query request - Primary gets a delete doc (id=X, v=2, s#=30)
4) Delete doc is processed first on the replica (id=X, v=2, s#=30)
5) Indexing operation arrives on the replica, since it's an auto-gen-id request and the retry marker is lower, we put it into lucene without any check. Replica has a doc the primary doesn't have.

To deal with a potential conflict between an append-only operation and a normal operation on replicas, we need to rely on sequence numbers. This commit maintains the max seqno of non-append-only operations on replica then only apply optimization for an append-only operation only if its seqno is higher than the seqno of all non-append-only.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -1032,6 +1038,35 @@ private boolean mayHaveBeenIndexedBefore(Index index) {
...
+         */
+        assert canOptimizeAddDocument(index);
+        assert index.origin() != Operation.Origin.PRIMARY;
+        return index.seqNo() <= maxSeqNoOfNonAppendOnlyOperations.get();
+    }
+
+    private void updateMaxSeqNoOfNonAppendOnlyOperations(Operation op) {
+        assert op.origin() != Operation.Origin.PRIMARY;
+        maxSeqNoOfNonAppendOnlyOperations.updateAndGet(curr -> Math.max(op.seqNo(), curr));
+        assert maxSeqNoOfNonAppendOnlyOperations.get() >= op.seqNo();
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: this needs a message

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -1032,6 +1038,35 @@ private boolean mayHaveBeenIndexedBefore(Index index) {
...
+         * If a replica has seen a non-append-only operation with a higher seqno than the seqno of an append-only, it may have seen
+         * the document of that append-only request. However, if the seqno of an append-only is higher than seqno of any non-append-only
+         * requests, we can assert the replica have not seen the document of that append-only request, thus we can apply optimization.
+         */
+        assert canOptimizeAddDocument(index);
+        assert index.origin() != Operation.Origin.PRIMARY;
+        return index.seqNo() <= maxSeqNoOfNonAppendOnlyOperations.get();
+    }
+
+    private void updateMaxSeqNoOfNonAppendOnlyOperations(Operation op) {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: maybe just inline this into the `planIndexingAsNonPrimary` method? I think that would be cleaner.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 3]
```diff
@@ -1032,6 +1038,35 @@ private boolean mayHaveBeenIndexedBefore(Index index) {
...
+         * As soon as an append-only request was indexed into the primary, it can be exposed to a search then users can issue
+         * a follow up operation on it. In rare cases, the follow up operation can be arrived and processed on a replica before
+         * the original append-only. In this case we can't simply proceed with the append only without consulting the version map.
+         * If a replica has seen a non-append-only operation with a higher seqno than the seqno of an append-only, it may have seen
+         * the document of that append-only request. However, if the seqno of an append-only is higher than seqno of any non-append-only
+         * requests, we can assert the replica have not seen the document of that append-only request, thus we can apply optimization.
+         */
+        assert canOptimizeAddDocument(index);
+        assert index.origin() != Operation.Origin.PRIMARY;
+        return index.seqNo() <= maxSeqNoOfNonAppendOnlyOperations.get();
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I think you should inline this into `planIndexingAsNonPrimary` then we don't need all the asserts

**Review by s1monw — APPROVED** [review]
I left some suggestions but LGTM otherwise

**Comment by DaveCTurner** [comment]
Thanks @dnhatn - I think we should agree on https://github.com/elastic/elasticsearch-formal-models/pull/28 before we proceed with this. In particular, it's not yet clear that this is the only problem in this area, and we haven't settled whether we think replicas should have a different version map that only considers seq#s (the document version numbers are now only of importance on the primary).

**Comment by dnhatn** [comment]
@DaveCTurner I understand your concern. Boaz and I discussed and agreed to start the implementation for this and #28790. We hoped that our reasoning for these would be fine. Certainly, we need to sync between the formal model and implementation.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 4]
```diff
@@ -364,15 +365,16 @@ public int fillSeqNoGaps(long primaryTerm) throws IOException {
...
-    private void updateMaxUnsafeAutoIdTimestampFromWriter(IndexWriter writer) {
-        long commitMaxUnsafeAutoIdTimestamp = Long.MIN_VALUE;
+    private void bootstrapAppendOnlyInfoFromWriter(IndexWriter writer) {
         for (Map.Entry<String, String> entry : writer.getLiveCommitData()) {
-            if (entry.getKey().equals(MAX_UNSAFE_AUTO_ID_TIMESTAMP_COMMIT_ID)) {
-                commitMaxUnsafeAutoIdTimestamp = Long.parseLong(entry.getValue());
-                break;
+            final String key = entry.getKey();
+            if (key.equals(MAX_UNSAFE_AUTO_ID_TIMESTAMP_COMMIT_ID)) {
+                maxUnsafeAutoIdTimestamp.set(Math.max(maxUnsafeAutoIdTimestamp.get(), Long.parseLong(entry.getValue())));
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: question - why the leniency with max?

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 5]
```diff
@@ -893,11 +895,24 @@ public IndexResult index(Index index) throws IOException {
...
+             * a follow-up operation on it. In rare cases, the follow up operation can be arrived and processed on a replica before
+             * the original append-only. In this case we can't simply proceed with the append only without consulting the version map.
+             * If a replica has seen a non-append-only operation with a higher seqno than the seqno of an append-only, it may have seen
+             * the document of that append-only request. However if the seqno of an append-only is higher than seqno of any non-append-only
+             * requests, we can assert the replica have not seen the document of that append-only request, thus we can apply optimization.
+             */
             assert index.version() == 1L : "can optimize on replicas but incoming version is [" + index.version() + "]";
             plan = IndexingStrategy.optimizedAppendOnly(index.seqNo());
         } else {
+            if (appendOnlyRequest == false) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: can we introduce a method similar to mayHaveBeenIndexedBefore that does the check and also updates the `maxSeqNoOfNonAppendOnlyOperations`? I think it's good to have both marker handling consistent.

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 6]
```diff
@@ -4572,4 +4572,92 @@ public void testStressUpdateSameDocWhileGettingIt() throws IOException, Interrup
...
+        long lookupTimes = 0L;
+        final LocalCheckpointTracker localCheckpointTracker = engine.getLocalCheckpointTracker();
+        final int initDocs = between(0, 10);
+        for (int i = 0; i < initDocs; i++) {
+            index(engine, i);
+            lookupTimes++;
+        }
+        // doc1 is delayed and arrived after a non-append-only op.
+        final long seqNoDoc1 = localCheckpointTracker.generateSeqNo();
+        Engine.IndexResult regularDoc = engine.index(replicaIndexForDoc(
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: can we also test a delete?

**Review by bleskes — CHANGES_REQUESTED** [review]
This is great. I left some nits. As discussed before, we should wait for blessing via @DaveCTurner 's work before merging.

**Comment by bleskes** [comment]
One more nit - can we maybe give the PR title to be more exact about what the code change? something like do not optimize if a normal operation was seen with a higher seq# ..

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 4 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I removed the max expr.

**Comment by dnhatn** [comment]
@bleskes I've addressed your comments. Would you please take another look? Thank you.

**Comment by DaveCTurner** [comment]
I have updated https://github.com/elastic/elasticsearch-formal-models/pull/28 as we discussed, so that's awaiting further comments. AIUI the model covers both this PR and #28790. From a first look, this seems to be quite different from the modelled solution and I'd be more comfortable if we brought them closer together (from either/both ends). In particular, we don't yet seem to be looking at the sequence numbers of replication requests here and I think we should.

I'm still travelling and haven't had a lot of sleep so expect more useful/correct/accurate feedback at a later date. I'm sending this now in case it's useful to those in more westerly timezones in the meantime.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 5 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Apparently @s1monw prefered the reverse. I'm fine with leaving as is.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 4 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Looks like the max is still here for time stamps? maybe assert it's -1 (for both seq# and timestamp)

**Review by bleskes — APPROVED** [review]

**Comment by bleskes** [comment]
> In particular, we don't yet seem to be looking at the sequence numbers of replication requests here and I think we should.

@DaveCTurner can you clarify what you mean? we use the seq# of the incoming op to update the `maxSeqNoOfNonAppendOnlyOperations` field?

**Comment by DaveCTurner** [comment]
Apologies, that was written on my phone and is inaccurate. I meant that the modelled solution stores seq#s in the version map.

**Review by DaveCTurner — APPROVED** [review]
LGTM.

**Comment by dnhatn** [comment]
Thanks @s1monw, @bleskes and @DaveCTurner for reviewing.

**Pull request merged**